cli/generate/src/render.rs

```diff
@@ -728,6 +740,73 @@ impl Generator {
         add_line!(self, "");
     }
 
+    fn add_supertype_map(&mut self) {
+        add_line!(
+            self,
+            "static const TSSymbol ts_supertypes[SUPERTYPE_COUNT] = {{"
+        );
+        indent!(self);
+        for supertype in self.supertype_map.keys() {
+            add_line!(self, "{supertype},");
+        }
+        dedent!(self);
+        add_line!(self, "}};\n");
+
+        add_line!(
+            self,
+            "static const TSSupertypeMapSlice ts_supertype_map_slices[] = {{",
+        );
+        indent!(self);
+        let mut row_id = 0;
+        let mut supertype_ids = vec![0];
+        let child_type_to_string = |ct: &ChildType| match ct {
+            ChildType::Normal(symbol) => self.string_for_symbol(symbol),
+            ChildType::Aliased(alias) => self.sanitize_string(&alias.value),
+        };
+        let mut supertype_string_map = BTreeMap::new();
+        for (supertype, subtypes) in &self.supertype_map {
+            supertype_string_map.insert(
+                supertype,
+                subtypes
+                    .iter()
+                    .map(child_type_to_string)
+                    .collect::<BTreeSet<String>>(),
+            );
+        }
```

Comment:
Continuing on with what I said about mapping symbol ids instead, here we wouldn't need the child type info to get the symbol string name, avoiding the need for this mapping of symbol or alias to the name.

Reply:
I had aimed for a supertype id -> subtype ids relationship initially and met some challenges. I might need your help with it, at the time it seemed even impossible. There are some cases (when aliases are involved) that we have subtypes aliased to an otherwise non-existant symbol name. Consider this rule in `tree-sitter-rust`:

```js
    _type: $ => choice(
      $.abstract_type,
      $.reference_type,
      $.metavariable,
      $.pointer_type,
      $.generic_type,
      $.scoped_type_identifier,
      $.tuple_type,
      $.unit_type,
      $.array_type,
      $.function_type,
      $._type_identifier,
      $.macro_invocation,
      $.never_type,
      $.dynamic_type,
      $.bounded_type,
      $.removed_trait_bound,
      alias(choice(...primitiveTypes), $.primitive_type),
    ),
```

here `_type` is a supertype. `primitiveTypes` is just a list of literals. `$.primitive_type` is not a real symbol, it only exists for this alias. In this case, from the parser.c side, there is no symbol ID for `primitive_type`. Thus what do we do here in this mapping? Surely we still want to record it somewhere since we can query `(primitive_type)`

Reply:
Ah, that's a good point...aliasing tokens is a little different. I think what we'd want here is to do one of two things:

- use the "base" token that the other tokens link back to, because they're all "hidden", in this case it'd be `anon_sym_u8` as it's just the first one
- add them all

I'm inclined to say we should add them all, as conceptually, we have n different tokens that are all aliased back to primitive type, and these tokens all have their symbol ids available for inspection

Reply:
Gotcha, that makes sense. I'm working on a solution that adds them all right now and it isn't as complicated as I thought initially